### PR TITLE
[BEAM-11659] Implement Pub/Sub Lite Table Provider

### PR DESCRIPTION
**Pull request by dpcollins-google**
Implement a TableProvider for Pub/Sub Lite, enabling it to be used in BeamSQL

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by dpcollins-google** [comment]
R: @amaliujia

**Comment by dpcollins-google** [comment]
Run SQL_Java11 PreCommit

**Comment by dpcollins-google** [comment]
Run SQL PreCommit

**Comment by dpcollins-google** [comment]
Run Java PreCommit

**Comment by dpcollins-google** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by dpcollins-google** [comment]
Run Java_Examples_Dataflow_Java11 PreCommit

**Comment by dpcollins-google** [comment]
Run SQL_Java11 PreCommit

**Comment by amaliujia** [comment]
cc @robinyqiu

**Comment by amaliujia** [comment]
I will try to start review this PR this week.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsublite/PubsubLiteTopicTable.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      TopicPath topic,
+      PTransform<PCollection<? extends Row>, PCollection<PubSubMessage>> transform) {
+    super(schema);
+    this.topic = topic;
+    this.transform = transform;
+  }
+
+  @Override
+  public PCollection<Row> buildIOReader(PBegin begin) {
+    throw new UnsupportedOperationException(
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsublite/PubsubLiteTopicTable.java:50]: Is there a test for this?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java [hunk 2]
```diff
@@ -949,6 +950,40 @@ public int hashCode() {
             getMetadata()
           });
     }
+
+    @Override
+    public String toString() {
```
Comment (amaliujia) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java:955]: I am not sure here: without this toString(), what is the default toString output for, say ROW type?

**Review by amaliujia — COMMENTED** [review]
Overall LGTM. Though this is a large PR and what this PR is adding is a isolated thing.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java (hunk 2 above)
Comment (dpcollins-google) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java:955]: This is FieldType, without it, its just a dump of all the fields in a json-ish structure, which makes it hard to determine at a glance what is going on in debug prints, and makes it not a great method to present to end users when saying "this field is badly formatted". This makes the output very similar to the DDL you would have to type to describe a field of this type.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsublite/PubsubLiteTopicTable.java (hunk 1 above)
Comment (dpcollins-google) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsublite/PubsubLiteTopicTable.java:50]: To clarify, you want me to test that you can't read from write tables or write from read tables? Added.

**Comment by dpcollins-google** [comment]
Run Java PreCommit

**Comment by dpcollins-google** [comment]
Run Java_Examples_Dataflow PreCommit

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsublite/PubsubLiteTopicTable.java (hunk 1 above)
Comment (amaliujia) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsublite/PubsubLiteTopicTable.java:50]: Yes. Thanks.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java (hunk 2 above)
Comment (amaliujia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java:955]: I see. In this case, do you think if it is worth also having simple tests for the new toString() code added (just for MAP,ROW,ARRAY)?

**Comment by amaliujia** [comment]
cc @robinyqiu 

I think this PR is in good shape (just have one open comment left). So if Robin has no other comments I can merge this PR after the open comment is addressed.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java (hunk 2 above)
Comment (amaliujia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java:955]: @dpcollins-google  what do you think of this comment?

Comment (dpcollins-google) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java:955]: Will add a test

Comment (dpcollins-google) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java:955]: Added

**Review by amaliujia — APPROVED** [review]
If there is no other comments, I will merge this PR tomorrow.

**Comment by amaliujia** [comment]
retest this please

**Comment by robinyqiu** [comment]
I took a quick look and this PR LGTM. We can merge after tests pass.

**Comment by amaliujia** [comment]
The failed tests seem are not flaky:

```
java.lang.NoSuchMethodError: 'void com.sun.tools.javac.util.Log.error(com.sun.tools.javac.util.JCDiagnostic$DiagnosticPosition, java.lang.String, java.lang.Object[])'
```

**Comment by dpcollins-google** [comment]
This looks to be failing when using java 11 with gradle? When using java 8 with gradle, :sdks:java:core:javadoc fails with confusing errors. Going to try rebasing

**Comment by dpcollins-google** [comment]
https://issues.apache.org/jira/browse/BEAM-11921

**Comment by amaliujia** [comment]
I am trying to triage BEAM-11921 to get it solved.

**Comment by amaliujia** [comment]
BEAM-11921 might be get fixed today.

**Comment by amaliujia** [comment]
@dpcollins-google can you rebase? https://github.com/apache/beam/pull/14148 should fix the failing tests.

**Pull request merged**